### PR TITLE
Introduce integration test source set for gradle based project

### LINKED ISSUE #22035: Gradle Plugin - Introduce integration test source set
This came from #21997 and #23488.

#21997 moves the `maven-failsafe-plugin` configuration out of the `native` profile and into the main profile.

#23488 changes the way codestarts generate sample code - specifically removing `@NativeImageTest` in favor of `@QuarkusIntegrationTest`.

When generating a Gradle project, the project contains a `native-test` source set. This name no longer aligns with the notion of integration test.

This issue is for the creation of a new source set representing integration tests (call it `int-test`, `intTest`, `integration-test`, or whatever else you'd like to call it).

This would all have to be backwards compatible of course, for a period of time. Projects already out there that have the `native-test` source set should still work, and any customizations to it should "flow" into this new source set.

It would have a rippling effect as well, since the Gradle plugin also has the `testNative` task, along with the `nativeTestImplementation` and `nativeTestRuntimeOnly` configurations.

To be consistent we'd have to introduce a new source set (something like `int-test`). We'd have to figure out a proper hierarchy (is the new source set the parent or child of the existing `native-test` source set?) and then apply configuration to it.

The new source set would most likely have to become the child of the existing one, so that any tweaks any current users have made will automagically "flow" into this new source set.

Then we'd have to update the documentation. And do we also introduce a new task `testIntegration` to be consistent? Now we have the same backwards compatibility problem there as well. We'd need to maintain backwards compatibility for a period of time.

The direction that is decided as part of #21997 (whether to always run integration test or have the user opt-into it - [this comment](https://github.com/quarkusio/quarkus/pull/21997#issuecomment-1035001857)) may dictate this behavior (i.e. keep it consistent). Personally I would think that it _should_ all be tied back into the `build` task, so running `./gradlew build` would run the unit tests, build the application (native, container image, jar, etc), and then execute the integration tests against whatever is built.

_Originally posted by @edeandrea in https://github.com/quarkusio/quarkus/issues/21997#issuecomment-988022858_

### LINKED ISSUE #23528: Allow `@QuarkusTest` and `@QuarkusIntegrationTest` to coexist in same test suite
### Description

I am running across some major hurtles with `@QuarkusTest` vs `QuarkusIntegrationTest`. Essentially, it makes sense to me as a developer that I would want to run both in the same `test*` command, and for the test codebase to comingle.

What I want to accomplish with this:

 - have the same tests for both native and non-native mode (possible in Maven, but not Gradle?)
 - simplify test writing, running, and configuration; nothing special for integration tests
 - potentially simplify Jacoco coverage? (Haven't looked into it, but I might think the extra config typically required for coverage of integration tests would be resolved with these proposed changes)

Currently though it seems that due to how the app is packaged for the different types of test, this is not currently supported and requires multiple code bases and different commands to be used to run them.

In Gradle, this seems to be the purpose of the `native-test` source set at the core, and in Maven an amount of pom configuration.

I suggest that we support a homogeneous codebase and command for running both `@QuarkuTest` and `@QuarkusIntegrationTest` tests. This would drastically simplify test creation and required configuration. If one wanted to have additional source-sets for Gradle runs it should still be possible, but one could also just run the same source set of `test` in native mode, and have the same tests available for jar integration testing.

I also realize I could be missing something and a reasonable solution might exist but I am not seeing it.

### Implementation ideas

We already have the ability to segment/ order tests by profiles and test resources; if we added the ability to order based on the two test annotations, we can separate the required runtime contexts and ensure they can run in the same suite/ run. It is possible I have a naive view of how the tests execution/ordering is handled, but I believe this is possible based on a couple of assumptions:

- we can carefully order the tests
- we have an amount of control of the running instance of quarkus used for each test

### PR DESCRIPTION
**Pull request by glefloch**
This introduce a new `intTest` source set (`src/intTest/java`) and a new `quarkusIntTest` task that execcute `@QuarkusIntegrationTest`.
This tasks depends on `quarkusBuild` to make sure the final artifact used by integration test is generated.

Next step will be: 
* Update code start to create the `intTest` sourceSet instead of `native-test`
* Update `nativeTest` task to combine both sourceSet
* Deprecate the `native-test` sourceSet

Close #23528
Close #22035

### CODE REVIEW

**Comment by glefloch** [comment]
cc @edeandrea

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestUtil.java [hunk 1]
```diff
@@ -498,7 +498,7 @@ private static Path determineBuildOutputDirectory(final URL url) {
             if (url.getPath().endsWith("test-classes/")) {
                 // we have the maven test classes dir
                 return toPath(url).getParent();
-            } else if (url.getPath().endsWith("test/")) {
+            } else if (url.getPath().endsWith("test/") || url.getPath().endsWith("intTest/")) {
```
Comment (aloubyansky) [inline test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestUtil.java:501]: We should check whether what we are doing if `url.getPath().endsWith("-tests.jar")` will work for all the other cases. That could be a separate PR though.

**Review by aloubyansky — APPROVED** [review]

**Review by geoand — APPROVED** [review]

**Comment by glefloch** [comment]
@edeandrea is it ok for you? this introduce a new `quarkusIntTest` task which will run tests located in `src/intTest/java`. 
The tasks depends on the artifact produced by `quarkusBuild`. 
The `quarkusIntTest` task is not linked to any other task.

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java [hunk 2]
```diff
@@ -244,6 +250,24 @@ public void execute(Task test) {
...
+                                    .plus(mainSourceSet.getOutput())
+                                    .plus(testSourceSet.getOutput()));
+
+                    tasks.register(INTEGRATION_TEST_TASK_NAME, Test.class, intTestTask -> {
+                        intTestTask.setGroup("verification");
+                        intTestTask.setDescription("Runs Quarkus integration tests");
+                        intTestTask.dependsOn(quarkusBuild);
+                        intTestTask.setClasspath(intTestSourceSet.getRuntimeClasspath());
+                        intTestTask.setTestClassesDirs(intTestSourceSet.getOutput().getClassesDirs());
+                    });
```
Comment (edeandrea) [inline devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java:269]: Do you need `intTestTask.setShouldRunAfter(List.of(tasks.findByName(JavaPlugin.TEST_TASK_NAME)));`?

File: docs/src/main/asciidoc/gradle-tooling.adoc [hunk 3]
```diff
@@ -501,6 +501,18 @@ quarkus {
...
+Quarkus integration tests (annotated with `@QuarkusIntegrationTest`) will run on the artifact produced by Quarkus.
+Those tests can be placed in a `src/intTest/java` directory and executed using:
+
+[source, bash]
+----
+./gradlew quarkusIntTest
+----
+
+This task depends on `quarkusBuild`. The final artifact will be produced before running tests.
+
```
Comment (edeandrea) [inline docs/src/main/asciidoc/gradle-tooling.adoc:515]: Does `quarkusBuild` also run the `check` task? If so, maybe you should mention that the unit tests will also be executed?

Maybe not necessary, but just a thought.

File: integration-tests/gradle/src/main/resources/it-test-basic-project/src/main/java/org/acme/ExampleResource.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package org.acme;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
```
Comment (edeandrea) [inline integration-tests/gradle/src/main/resources/it-test-basic-project/src/main/java/org/acme/ExampleResource.java:9]: I don't think this import is necessary?

File: integration-tests/gradle/src/main/resources/it-test-basic-project/src/test/java/org/acme/ExampleResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+@QuarkusTest
+public class ExampleResourceTest {
+
+    @Test
+    public void testHelloEndpoint() {
+        given()
+          .when().get("/hello")
+          .then()
+             .statusCode(200)
+             .body(is("Hello from Test"));
```
Comment (edeandrea) [inline integration-tests/gradle/src/main/resources/it-test-basic-project/src/test/java/org/acme/ExampleResourceTest.java:18]: I don't understand how this test passes? `ExampleResource` only has a single endpoint, `/hello`, which returns `"Hello World"`?

File: integration-tests/gradle/src/test/java/io/quarkus/gradle/IntegrationTestBuildTest.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+import org.junit.jupiter.api.Test;
+
+public class IntegrationTestBuildTest extends QuarkusGradleWrapperTestBase {
+
+    @Test
+    public void shouldRunIntegrationTestAsPartOfBuild() throws Exception {
+        File projectDir = getProjectDir("it-test-basic-project");
+
+        BuildResult buildResult = runGradleWrapper(projectDir, "clean", "quarkusIntTest");
```
Comment (edeandrea) [inline integration-tests/gradle/src/test/java/io/quarkus/gradle/IntegrationTestBuildTest.java:15]: See comment above re: should the `intTest` task trigger the `check` (or at a minimum `test`) task

**Comment by edeandrea** [comment]
@glefloch I added some comments.

Also, should this also close #22035?

File: docs/src/main/asciidoc/gradle-tooling.adoc (hunk 3 above)
Comment (glefloch) [reply docs/src/main/asciidoc/gradle-tooling.adoc:515]: Currently `quarkusBuild` is not linked to any `test` task

Comment (edeandrea) [reply docs/src/main/asciidoc/gradle-tooling.adoc:515]: Should it (or the `intTest` task) be?

File: integration-tests/gradle/src/main/resources/it-test-basic-project/src/test/java/org/acme/ExampleResourceTest.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    @Test
+    public void testTestOnly() {
+      given()
+          .when().get("test-only")
+          .then()
+              .statusCode(200)
+              .body(is("Test only"));
```
[the remark this answers is not in the document]
Comment (glefloch) [reply integration-tests/gradle/src/main/resources/it-test-basic-project/src/test/java/org/acme/ExampleResourceTest.java:27]: I removed it.

File: integration-tests/gradle/src/test/java/io/quarkus/gradle/IntegrationTestBuildTest.java (hunk 6 above)
Comment (glefloch) [reply integration-tests/gradle/src/test/java/io/quarkus/gradle/IntegrationTestBuildTest.java:15]: Right, I will add a dependency the `check` task.

File: docs/src/main/asciidoc/gradle-tooling.adoc (hunk 3 above)
Comment (glefloch) [reply docs/src/main/asciidoc/gradle-tooling.adoc:515]: I will add a dependency between `quarkusIntTest` and `check`

**Comment by gsmet** [comment]
Is `intTest` used by other projects? It sounds very weird to me. Maybe `it` would be better? True question :).

**Comment by edeandrea** [comment]
Are you talking about the source set name? Personally I think `it` is too short. I think the test task name is `quarkusIntTest`.

**Comment by glefloch** [comment]
There isn't convention for this. 
In documention they usually declare ˋsrc/integrationTest/java` 
And use ˋintegrationTest` task name. Maybe we should go with that to be as close as possible from docs.

**Comment by edeandrea** [comment]
`IntegrationTest` is just so long and wordy. My last 2 companies we used `intTest` and the test task name was also `intTest`, following the convention that the test task name matches the source set name.

**Comment by gsmet** [comment]
Mkay. Personally, I never saw `intTest` being used but I'll let you guys do what you prefer given I don't use Gradle :).

**Comment by edeandrea** [comment]
Actually I was just looking at one of my personal projects. I use `intTest` as the source set name (https://github.com/edeandrea/xjc-generation-gradle-plugin/blob/main/build.gradle.kts#L27) but `integrationTest` as the task name (https://github.com/edeandrea/xjc-generation-gradle-plugin/blob/main/build.gradle.kts#L100). I also make the `integrationTest` task run after the `test` task (https://github.com/edeandrea/xjc-generation-gradle-plugin/blob/main/build.gradle.kts#L105) & the `check` task depend on the `integrationTest` task (https://github.com/edeandrea/xjc-generation-gradle-plugin/blob/main/build.gradle.kts#L114).

@gsmet you should use it - it's a far better tool IMO :)

**Comment by glefloch** [comment]
I updated the task name to match the name of the source set. 
Looking at gradle documentation, there are both `intTest` and `integrationTest`. 

`integrationTest` is longer but may be more relevant that `intTest`. `it` works too.. 

@aloubyansky, @geoand do you have a preference ?

**Comment by geoand** [comment]
I vote for `integrationTest`

**Comment by aloubyansky** [comment]
Another thing is, so far, every command in the Quarkus plugin had `quarkus` prefix. Could `integrationTest` potentially conflict with other plugins? Of course `quarkusInterationTest` will make it even longer, which isn't great.

**Comment by glefloch** [comment]
Yes this is a risk. 

We can make the source set name configurable with a default value of ˋintegrationTest` and use ˋquarkusIntTest` as task name ?

**Comment by edeandrea** [comment]
> Another thing is, so far, every command in the Quarkus plugin had `quarkus` prefix. Could `integrationTest` potentially conflict with other plugins? Of course `quarkusInterationTest` will make it even longer, which isn't great.

I agree - keep it consistent. If every task the plugin introduces has the `quarkus` prefix then this probably should too. `quarkusIntTest` works for me.

Personally I like `intTest` over `integrationTest` as the source set name, but at the end of the day it doesn't really matter that much to me. If you all like `integrationTest` then so be it.

**Comment by glefloch** [comment]
I adressed all comments, updated the sourceSet to `integrationTest` and the task name to `quarkusIntTest`

**Comment by edeandrea** [comment]
Nice work!

**Comment by edeandrea** [comment]
Hey @glefloch should the [`Using @QuarkusIntegrationTest` section in the getting started testing guide](https://quarkus.io/guides/getting-started-testing#quarkus-integration-test) include something about this new `quarkusIntTest` task?

Currently it just says

> As a test annotated with `@QuarkusIntegrationTest` tests the result of the build, it should be run as part of the integration test suite - i.e. via the `maven-failsafe-plugin` if using Maven or an additional task if using Gradle. These tests will not work if run in the same phase as `@QuarkusTest` as Quarkus has not yet created the final artifact.

Maybe it should say something like

As a test annotated with `@QuarkusIntegrationTest` tests the result of the build, it should be run as part of the integration test suite - i.e. via the `maven-failsafe-plugin` if using Maven or the `quarkusIntTest` task if using Gradle. These tests will not work if run in the same phase as `@QuarkusTest` as Quarkus has not yet created the final artifact.

?

**Comment by glefloch** [comment]
Right, I will add a note on this guide.

**Comment by rinmalavi** [comment]
> Right, I will add a note on this guide.

Not sure if ' or the quarkusIntTest task if using Gradle.' is the note added.
Reading this discussion I would assume `quarkusIntTest` will run only `@QuarkusIntegrationTest` in `integrationTest` source set, but it runs everything. 
I am guessing i should add something to gradle.build to make this work, or was this supposed to be out of the box.
